File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int res=Integer.MIN_VALUE;
    public int MaxPath(TreeNode root){
        if(root==null) return 0;
        int l=MaxPath(root.left);
        int r=MaxPath(root.right);
        int t=root.val+Math.max(0,Math.max(l,r));
        int mx=Math.max(t,root.val+l+r);
        res=Math.max(mx,res);
        return t;
    }
    public int maxPathSum(TreeNode root) {
        MaxPath(root);
        return res;
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int MaxPath(TreeNode* root, int& res){
        if(root){
            int l=MaxPath(root->left,res);
            int r=MaxPath(root->right,res);
            int mx=root->val+max(0,max(l,r));
            int Max=max(mx,root->val+l+r);
            res=max(Max,res);
            return mx;
        }
        return 0;
    }
    int maxPathSum(TreeNode* root) {
        int res=INT_MIN;
        MaxPath(root,res);
        return res;
    }
};